### DISCUSSION TITLE
Carbon API Usage

### DISCUSSION THREAD

**Discussion by dzoidbrg** [FelixKratz/SketchyBar#381] [Q&A] [answered]
Hi! I am a Linux user, who also occasionally uses MacOS with yabai and SketchyBar. I was looking through the code, and I stumbled upon #include <Carbon/Carbon.h>. Isn't the Carbon API deprecated? Why does this project need it? 

Any answers would be appreciated,

Jakob

**Comment by FelixKratz** [comment, marked as answer]
I have removed some of the Carbon includes on master, since they where not necessary: 5b820d3b5d2ae881481d7275efa11750353b8763

However, for some things I still use them (while there might be alternatives):
* Mouse Event Handling uses the symbol: `InstallEventHandler`
* The private symbol `RunApplicationEventLoop` is used to run the application run loop (there is a specific reason why I use this instead of simple `CFRunLoopRun`)

You will notice that yabai also uses some Carbon functionality: https://github.com/koekeishiya/yabai/blob/df94ce05a297c1e62ff1eaa1f8354fcf2193126c/src/manifest.m#L1

**Reply by dzoidbrg** [reply]
Yes, I also saw Yabai using it! Isn't its deprecation a problem?

**Reply by FelixKratz** [reply]
> Yes, I also saw Yabai using it! Isn't its deprecation a problem?

I don't think it is a problem, it is still widely used internally. However, on this branch I have made the changes necessary to entirely remove Carbon from this project:
https://github.com/FelixKratz/SketchyBar/compare/master...remove_carbon

The events are directly received on the mach port and pulled from the WindowServer.

**Reply by dzoidbrg** [reply]
Thank you for your answer. Last question, are any of these mach APIs documented? I am trying to write a MacOS keyboard shortcut tool which is fit for my usecase,

**Reply by dzoidbrg** [reply]
Thank you so much for your incredibly detailed reply. Sorry for troubling you further with this question. How do you link SkyLight.framework. Isn't it a PrivateFramework in /System/PrivateFrameworks which requires SIP to be disabled to link? Or did I misunderstand this.

**Reply by FelixKratz** [reply]
You can link to SkyLight on any macOS System, it has nothing to do with SIP. Simply add these flags: `-F/System/Library/PrivateFrameworks -framework SkyLight`

Alternatively you can use `dlopen` and `dlsym` to get the offsets of those private functions.
